File: mpga-plugin/cli/src/generators/index-md.ts
import fs from 'fs';
import path from 'path';
import { ScanResult, detectProjectType, getTopLanguage } from '../core/scanner.js';
import { MpgaConfig } from '../core/config.js';
import { ScopeInfo } from './scope-md.js';

export function renderIndexMd(
  scanResult: ScanResult,
  config: MpgaConfig,
  scopes: ScopeInfo[],
  activeMilestone: string | null,
  evidenceCoverage: number
): string {
  const now = new Date().toISOString();
  const projectType = detectProjectType(scanResult);
  const { totalFiles, totalLines, languages } = scanResult;

  const langSummary = Object.entries(languages)
    .sort((a, b) => b[1].lines - a[1].lines)
    .map(([lang, stats]) => `${lang} (${Math.round((stats.lines / totalLines) * 100)}%)`)
    .join(', ');

  const lines: string[] = [];

  lines.push(`# Project: ${config.project.name}`, '');
  lines.push('## Identity');
  lines.push(`- **Type:** ${projectType}`);
  lines.push(`- **Size:** ~${totalLines.toLocaleString()} lines across ${totalFiles} files`);
  lines.push(`- **Languages:** ${langSummary}`);
  lines.push(`- **Last sync:** ${now}`);
  lines.push(`- **Evidence coverage:** ${Math.round(evidenceCoverage * 100)}% (target: ${Math.round(config.evidence.coverageThreshold * 100)}%)`);
  lines.push('');

  // Key files table — top 10 by size
  lines.push('## Key files');
  lines.push('| File | Role | Evidence |');
  lines.push('|------|------|----------|');
  const roleMap = config.knowledgeLayer?.keyFileRoles ?? {};
  const topFiles = [...scanResult.files].sort((a, b) => b.lines - a.lines).slice(0, 10);
  for (const f of topFiles) {
    const role = roleMap[f.filepath] ?? '(describe role)';
    lines.push(`| ${f.filepath} | ${role} | [E] ${f.filepath}:1-${Math.min(50, f.lines)} |`);
  }
  lines.push('');

  lines.push('## Conventions');
  const customConventions = config.knowledgeLayer?.conventions?.filter((c) => c.trim().length > 0);
  if (customConventions && customConventions.length > 0) {
    for (const c of customConventions) {
      lines.push(`- ${c}`);
    }
  } else {
    lines.push('- (Add your project conventions here)');
    lines.push('- (e.g. "All API routes follow REST naming: /api/v1/<resource>")');
  }
  lines.push('');

  lines.push('## Agent trigger table');
  lines.push('| Task pattern | Agent | Scopes to load |');
  lines.push('|-------------|-------|-----------------|');
  lines.push('| "add/modify authentication" | green-dev → red-dev → blue-dev | auth, database |');
  lines.push('| "explore how X works" | scout | (auto-detect) |');
  lines.push('| "plan feature X" | researcher → architect | (auto-detect) |');
  lines.push('| "fix bug in X" | scout → green-dev → red-dev | (auto-detect) |');
  lines.push('| "refactor X" | architect → blue-dev | (auto-detect) |');
  lines.push('');

  lines.push('## Scope registry');
  lines.push('| Scope | Status | Evidence links | Last verified |');
  lines.push('|-------|--------|---------------|---------------|');
  const today = now.split('T')[0];
  for (const scope of scopes) {
    lines.push(`| ${scope.name} | ✓ fresh | 0/${scope.exports.length} | ${today} |`);
  }
  lines.push('');

  lines.push('## Active milestone');
  lines.push(`- ${activeMilestone ?? '(none)'}`);
  lines.push('');

  lines.push('## Known unknowns');
  lines.push('- [ ] (run `mpga evidence verify` to detect unknowns)');

  return lines.join('\n') + '\n';
}

File: mpga-plugin/cli/src/mpga/generators/index_md.py
from __future__ import annotations

from datetime import UTC, datetime

from mpga.core.config import MpgaConfig
from mpga.core.scanner import ScanResult, detect_project_type
from mpga.generators.scope_md import ScopeInfo

# Number of top files (by line count) to list in the Key Files table.
TOP_FILES_COUNT = 10
# Maximum lines to reference in an evidence link for a key file.
KEY_FILE_EVIDENCE_MAX_LINES = 50


def render_index_md(
    scan_result: ScanResult,
    config: MpgaConfig,
    scopes: list[ScopeInfo],
    active_milestone: str | None,
    evidence_coverage: float,
    scope_health: dict | None = None,
) -> str:
    now = datetime.now(UTC).isoformat()
    project_type = detect_project_type(scan_result)
    total_files = scan_result.total_files
    total_lines = scan_result.total_lines
    languages = scan_result.languages

    lang_summary = ", ".join(
        f"{lang} ({round(stats['lines'] / total_lines * 100)}%)"
        for lang, stats in sorted(
            languages.items(), key=lambda item: item[1]["lines"], reverse=True
        )
    )

    lines: list[str] = []

    lines.append(f"# Project: {config.project.name}")
    lines.append("")
    lines.append("## Identity")
    lines.append(f"- **Type:** {project_type}")
    lines.append(
        f"- **Size:** ~{total_lines:,} lines across {total_files} files"
    )
    lines.append(f"- **Languages:** {lang_summary}")
    lines.append(f"- **Last sync:** {now}")
    lines.append(
        f"- **Evidence coverage:** {round(evidence_coverage * 100)}%"
        f" (target: {round(config.evidence.coverage_threshold * 100)}%)"
    )
    lines.append("")

    # Key files table -- top 10 by size
    lines.append("## Key files")
    lines.append("| File | Role | Evidence |")
    lines.append("|------|------|----------|")
    role_map: dict[str, str] = {}
    if config.knowledge_layer is not None:
        role_map = config.knowledge_layer.key_file_roles
    top_files = sorted(scan_result.files, key=lambda f: f.lines, reverse=True)[
        :TOP_FILES_COUNT
    ]
    for f in top_files:
        role = role_map.get(f.filepath, "(describe role)")
        lines.append(
            f"| {f.filepath} | {role}"
            f" | [E] {f.filepath}:1-{min(KEY_FILE_EVIDENCE_MAX_LINES, f.lines)} |"
        )
    lines.append("")

    lines.append("## Conventions")
    custom_conventions: list[str] = []
    if config.knowledge_layer is not None:
        custom_conventions = [
            c for c in config.knowledge_layer.conventions if c.strip()
        ]
    if custom_conventions:
        for c in custom_conventions:
            lines.append(f"- {c}")
    else:
        lines.append("- (Add your project conventions here)")
        lines.append('- (e.g. "All API routes follow REST naming: /api/v1/<resource>")')
    lines.append("")

    lines.append("## Agent trigger table")
    lines.append("| Task pattern | Agent | Scopes to load |")
    lines.append("|-------------|-------|-----------------|")
    lines.append(
        '| "add/modify authentication" | red-dev \u2192 green-dev \u2192 blue-dev | auth, database |'
    )
    lines.append('| "explore how X works" | scout | (auto-detect) |')
    lines.append(
        '| "plan feature X" | researcher \u2192 architect | (auto-detect) |'
    )
    lines.append(
        '| "fix bug in X" | scout \u2192 red-dev \u2192 green-dev | (auto-detect) |'
    )
    lines.append('| "refactor X" | architect \u2192 blue-dev | (auto-detect) |')
    lines.append('| "design/wireframe feature X" | designer | (auto-detect) |')
    lines.append('| "audit UI of X" | ui-auditor | (auto-detect) |')
    lines.append('| "check visual regression" | visual-tester | (auto-detect) |')
    lines.append('| "create design system" | designer \u2192 ui-auditor | (auto-detect) |')
    lines.append('| "frontend design for X" | mpga-frontend-design | (auto-detect) |')
    lines.append("")

    lines.append("## Scope registry")
    lines.append("| Scope | Status | Evidence links | Last verified |")
    lines.append("|-------|--------|---------------|---------------|")
    today = now.split("T")[0]
    for scope in scopes:
        if scope_health and scope.name in scope_health:
            sr = scope_health[scope.name]
            total = sr.total_links
            valid = sr.valid_links + sr.healed_links
            pct = sr.health_pct
            if pct >= 80:
                status = "✓ fresh"
            elif pct >= 50:
                status = "⚠ stale"
            else:
                status = "✗ broken"
            link_count = f"{valid}/{total}"
        else:
            status = "[Unknown]"
            link_count = f"0/{len(scope.exports)}"
        lines.append(f"| {scope.name} | {status} | {link_count} | {today} |")
    lines.append("")

    lines.append("## Active milestone")
    lines.append(f"- {active_milestone if active_milestone is not None else '(none)'}")
    lines.append("")

    lines.append("## Known unknowns")
    lines.append("- [ ] (run `mpga evidence verify` to detect unknowns)")

    return "\n".join(lines) + "\n"
